File: bindings/swift/Sources/WhatWG/Providers/TimerProvider.swift
import Foundation

/// Protocol for providing timer functionality.
///
/// Implement this protocol to provide setTimeout/setInterval support.
/// The default implementation uses GCD (Grand Central Dispatch).
///
public protocol TimerProvider: AnyObject, Sendable {
    
    /// Schedules a one-time callback.
    ///
    /// - Parameters:
    ///   - delay: Delay in milliseconds.
    ///   - callback: The callback to invoke.
    /// - Returns: A timer ID that can be used to cancel.
    func setTimeout(delay: UInt32, callback: @escaping @Sendable () -> Void) -> UInt64
    
    /// Schedules a repeating callback.
    ///
    /// - Parameters:
    ///   - interval: Interval in milliseconds.
    ///   - callback: The callback to invoke.
    /// - Returns: A timer ID that can be used to cancel.
    func setInterval(interval: UInt32, callback: @escaping @Sendable () -> Void) -> UInt64
    
    /// Cancels a timeout.
    ///
    /// - Parameter id: The timer ID returned by `setTimeout`.
    func clearTimeout(id: UInt64)
    
    /// Cancels an interval.
    ///
    /// - Parameter id: The timer ID returned by `setInterval`.
    func clearInterval(id: UInt64)
}

File: bindings/kotlin/whatwg/src/main/kotlin/com/whatwg/providers/TimerProvider.kt
package com.whatwg.providers

/**
 * Interface for providing timer functionality.
 *
 * Implement this interface to provide setTimeout/setInterval support.
 */
interface TimerProvider {

    /**
     * Schedules a one-time callback.
     *
     * @param delayMs Delay in milliseconds.
     * @param callback The callback to invoke.
     * @return A timer ID that can be used to cancel.
     */
    fun setTimeout(delayMs: Long, callback: () -> Unit): Long

    /**
     * Schedules a repeating callback.
     *
     * @param intervalMs Interval in milliseconds.
     * @param callback The callback to invoke.
     * @return A timer ID that can be used to cancel.
     */
    fun setInterval(intervalMs: Long, callback: () -> Unit): Long

    /**
     * Cancels a timeout.
     *
     * @param id The timer ID returned by [setTimeout].
     */
    fun clearTimeout(id: Long)

    /**
     * Cancels an interval.
     *
     * @param id The timer ID returned by [setInterval].
     */
    fun clearInterval(id: Long)
}
